source/fuzz/transformation_push_id_through_variable.cpp

```diff
@@ -0,0 +1,174 @@
+  message_.set_pointer_type_id(pointer_type_id);
+  message_.set_variable_storage_class(variable_storage_class);
+  *message_.mutable_instruction_descriptor() = instruction_descriptor;
+}
+
+bool TransformationPushIdThroughVariable::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  // |message_.value_synonym_id| and |message_.variable_id| must be fresh.
+  if (!fuzzerutil::IsFreshId(ir_context, message_.value_synonym_id()) ||
+      !fuzzerutil::IsFreshId(ir_context, message_.variable_id())) {
+    return false;
+  }
+
+  // The instruction to insert before must be defined.
+  auto instruction_to_insert_before =
+      FindInstruction(message_.instruction_descriptor(), ir_context);
+  if (!instruction_to_insert_before) {
+    return false;
+  }
+
+  // The instruction to insert before must belong to a reachable block.
+  auto basic_block = ir_context->get_instr_block(instruction_to_insert_before);
+  if (!fuzzerutil::BlockIsReachableInItsFunction(ir_context, basic_block)) {
+    return false;
+  }
+
+  // It must be valid to insert the OpStore and OpLoad instruction before it.
+  if (!fuzzerutil::CanInsertOpcodeBeforeInstruction(
+          SpvOpStore, instruction_to_insert_before) ||
+      !fuzzerutil::CanInsertOpcodeBeforeInstruction(
+          SpvOpLoad, instruction_to_insert_before)) {
+    return false;
+  }
+
+  // The value instruction must be defined and have a type.
+  auto value_instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.value_id());
+  if (!value_instruction || !value_instruction->type_id()) {
+    return false;
+  }
+
+  // The pointer type instruction must be defined and be an OpTypePointer
+  // instruction.
+  auto pointer_type_instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.pointer_type_id());
+  if (!pointer_type_instruction ||
+      pointer_type_instruction->opcode() != SpvOpTypePointer) {
+    return false;
+  }
+
+  // The pointer type storage class must be equal to the variable storage class.
+  if (pointer_type_instruction->GetSingleWordInOperand(0) !=
+      message_.variable_storage_class()) {
+    return false;
+  }
+
+  // The pointee type must be equal to the value type.
+  if (pointer_type_instruction->GetSingleWordInOperand(1) !=
+      value_instruction->type_id()) {
+    return false;
+  }
+
+  // |message_.variable_storage_class| must be private or function.
+  if (message_.variable_storage_class() != SpvStorageClassPrivate &&
+      message_.variable_storage_class() != SpvStorageClassFunction) {
+    return false;
+  }
+
+  // |message_.value_id| must be available at the insertion point.
+  return fuzzerutil::IdIsAvailableBeforeInstruction(
+      ir_context, instruction_to_insert_before, message_.value_id());
+}
+
+void TransformationPushIdThroughVariable::Apply(
+    opt::IRContext* ir_context,
+    TransformationContext* transformation_context) const {
+  // Adds whether a global or local variable.
```

Comment:
Instead of using pointer_type_id, you'll need to use the helper function I've suggested above to get that id.

Reply:
Done.